zallet/src/components/sync.rs

```diff
@@ -1,45 +1,323 @@
+            }
+        }
 
-            loop {
-                // TODO: Move this inside `sync::run` so that we aren't querying subtree roots
-                // every interval.
+        // If the fork point is equal to `prev_tip` then no reorg has occurred.
+        if fork_point != prev_tip {
+            // Ensured by `find_fork`.
+            assert!(fork_point.height < prev_tip.height);
+
+            // Rewind the wallet to the fork point.
+            // TODO: Is there anything else we should do with the blocks in the old fork?
+            info!(
+                "Chain reorg detected, rewinding to {} {}",
+                fork_point.height, fork_point.hash
+            );
+            db_data.truncate_to_height(fork_point.height)?;
+        }
+
+        // Notify the wallet of block connections.
+        db_data.update_chain_tip(current_tip.height)?;
+        if !block_stack.is_empty() {
+            let from_height =
+                BlockHeight::from_u32(block_stack.last().expect("not empty").height as u32);
+            let end_height =
+                BlockHeight::from_u32(block_stack.first().expect("not empty").height as u32 + 1);
+            let scan_range = ScanRange::from_parts(from_height..end_height, ScanPriority::ChainTip);
+            db_cache.insert(block_stack).await?;
+
+            let from_state = steps::fetch_chain_state(chain, from_height.saturating_sub(1)).await?;
+
+            tokio::task::block_in_place(|| {
+                info!("Scanning {}", scan_range);
+                scan_cached_blocks(
+                    params,
+                    &db_cache,
+                    db_data,
+                    from_height,
+                    &from_state,
+                    scan_range.len(),
+                )
+            })?;
+
+            db_cache.delete(scan_range).await?;
+        }
+
+        // Now that we're done applying the chain diff, update our chain pointers.
+        prev_tip = current_tip;
+        current_tip = steps::get_chain_tip(chain).await?;
+
+        // If the chain tip no longer matches, we have more to do before consuming mempool
+        // updates.
+        if prev_tip != current_tip {
+            continue;
+        }
+
+        // We have caught up to the chain tip. Stream the mempool state into the wallet.
+        info!("Reached chain tip, streaming mempool");
+        let mempool_height = current_tip.height + 1;
+        let consensus_branch_id = consensus::BranchId::for_height(params, mempool_height);
+        let mut mempool_stream = chain.get_mempool_stream().await?;
+        while let Some(result) = mempool_stream.next().await {
+            match result {
+                Ok(raw_tx) => {
+                    let tx = Transaction::read(
+                        SerializedTransaction::from(raw_tx.data).as_ref(),
+                        consensus_branch_id,
+                    )
+                    .expect("Zaino should only provide valid transactions");
```

Comment:
This should not be a panic. It might be okay to exit but the exit shouldn't be here, it should propagate as a SyncError.

Reply:
It is a programming error for Zaino to return an invalid transaction from the mempool stream. Really this is a problem with Zaino not having a typesafe Rust API.